Rest API doesn't connect to the created document store (?)

The title for this issue may be wrong, the issue might not be that at all..

Related to #4364 and again flagged by @rolandtannous and @recrudesce

With the same pipeline that is used in the issue #4364 another problem that was noticed:
When we initially start the pipeline the first thing we do is use the `file-upload` endpoint to write files to the document store. This works fine and we can see the embeddings being created and written in the logs. 

If we stop the rest API and re-start it with the same document store, the expected behaviour is that we simply connect to and use the embeddings and documents that were written to that document store before. However, this is not the behaviour that we observe and here are the logs that we receive:

(Below I will paste the YAML again for convenience) 

```
Calculating embeddings: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:28<00:00, 28.76s/it]
03/09/2023 11:05:56 AM Index is empty. First add some documents to search them.
03/09/2023 11:05:56 AM No documents with embeddings. Run the document store's update_embeddings() method.
[2023-03-09 11:05:56 +0100] [50290] [ERROR] Exception in ASGI application
Traceback (most recent call last):
  File "/Users/tuanacelik/Documents/git/haystack/haystack/pipelines/base.py", line 555, in run
    node_output, stream_id = self._run_node(node_id, node_input)
  File "/Users/tuanacelik/Documents/git/haystack/haystack/pipelines/base.py", line 469, in _run_node
    return self.graph.nodes[node_id]["component"]._dispatch_run(**node_input)
  File "/Users/tuanacelik/Documents/git/haystack/haystack/nodes/base.py", line 204, in _dispatch_run
    return self._dispatch_run_general(self.run, **kwargs)
  File "/Users/tuanacelik/Documents/git/haystack/haystack/nodes/base.py", line 248, in _dispatch_run_general
    output, stream = run_method(**run_inputs, **run_params)
  File "/Users/tuanacelik/Documents/git/haystack/haystack/nodes/prompt/prompt_node.py", line 663, in run
    results = self(prompt_collector=prompt_collector, **invocation_context)
  File "/Users/tuanacelik/Documents/git/haystack/haystack/nodes/prompt/prompt_node.py", line 475, in __call__
    return self.prompt(self.default_prompt_template, *args, **kwargs)
  File "/Users/tuanacelik/Documents/git/haystack/haystack/nodes/prompt/prompt_node.py", line 509, in prompt
    for prompt in template_to_fill.fill(*args, **kwargs):
  File "/Users/tuanacelik/Documents/git/haystack/haystack/nodes/prompt/prompt_node.py", line 150, in fill
    template_dict = self.prepare(*args, **kwargs)
  File "/Users/tuanacelik/Documents/git/haystack/haystack/nodes/prompt/prompt_node.py", line 129, in prepare
    raise ValueError(f"Expected prompt parameters {self.prompt_params} but got {list(available_params)}.")
ValueError: Expected prompt parameters ['documents', 'query'] but got ['top_k', 'query', 'stop_words'].
```

Example YAML:
```
version: ignore

components:    # define all the building-blocks for Pipeline
  - name: DocumentStore
    type: ElasticsearchDocumentStore
    params:
      host: localhost
      embedding_dim: 1536
      recreate_index: True
  - name: Retriever
    type: EmbeddingRetriever
    params:
      api_key: API-KEY
      document_store: DocumentStore
      embedding_model: text-embedding-ada-002
      top_k: 6
  - name: lfqa_template
    params:
      name: lfqa
      prompt_text: "Given the context please generate a comprehensive, summarized answer. Context: $documents; Question: $query; Answer:"
    type: PromptTemplate
  - name: Prompt
    type: PromptNode
    params:
      api_key: API-KEY
      model_name_or_path: text-davinci-003
      default_prompt_template: lfqa_template
      output_variable: answers
  - name: FileTypeClassifier
    type: FileTypeClassifier
  - name: TextFileConverter
    type: TextConverter
  - name: Preprocessor
    type: PreProcessor
    params:
      split_by: word
      split_length: 180
      split_overlap: 40
      split_respect_sentence_boundary: False

pipelines:
  - name: query    # a sample extractive-qa Pipeline
    nodes:
      - name: Retriever
        inputs: [Query]
      - name: Prompt
        inputs: [Retriever]
  - name: indexing
    nodes:
      - name: FileTypeClassifier
        inputs: [File]
      - name: TextFileConverter
        inputs: [FileTypeClassifier.output_1]
      - name: Preprocessor
        inputs: [TextFileConverter]
      - name: Retriever
        inputs: [Preprocessor]
      - name: DocumentStore
        inputs: [Retriever]
```

--- a/haystack/nodes/prompt/prompt_template.py
+++ b/haystack/nodes/prompt/prompt_template.py
@@ -268,8 +268,8 @@ class PromptTemplate(BasePromptTemplate, ABC):
                 if param in kwargs:
                     params_dict[param] = kwargs[param]
 
-        if not set(self.prompt_params).issubset(set(params_dict.keys())):
-            available_params = set(list(params_dict.keys()) + list(set(kwargs.keys())))
+        if not set(self.prompt_params).issubset(params_dict.keys()):
+            available_params = {*params_dict.keys(), *kwargs.keys()}
             provided = set(self.prompt_params).intersection(available_params)
             message = f"only {list(provided)}" if provided else "none of these parameters"
             raise ValueError(

--- a/haystack/nodes/prompt/prompt_template.py
+++ b/haystack/nodes/prompt/prompt_template.py
@@ -274,7 +274,7 @@ class PromptTemplate(BasePromptTemplate, ABC):
             message = f"only {list(provided)}" if provided else "none of these parameters"
             raise ValueError(
                 f"Expected prompt parameters {self.prompt_params} to be provided but got "
-                f"{message}. Please make sure to provide all template parameters."
+                f"{message}. Make sure to provide all template parameters."
             )
 
         template_dict = {"_at_least_one_prompt": True}

--- a/haystack/nodes/prompt/prompt_template.py
+++ b/haystack/nodes/prompt/prompt_template.py
@@ -268,9 +268,14 @@ class PromptTemplate(BasePromptTemplate, ABC):
                 if param in kwargs:
                     params_dict[param] = kwargs[param]
 
-        if set(params_dict.keys()) != set(self.prompt_params):
+        if not set(self.prompt_params).issubset(set(params_dict.keys())):
             available_params = set(list(params_dict.keys()) + list(set(kwargs.keys())))
-            raise ValueError(f"Expected prompt parameters {self.prompt_params} but got {list(available_params)}.")
+            provided = set(self.prompt_params).intersection(available_params)
+            message = f"only {list(provided)}" if provided else "none of these parameters"
+            raise ValueError(
+                f"Expected prompt parameters {self.prompt_params} to be provided but got "
+                f"{message}. Please make sure to provide all template parameters."
+            )
 
         template_dict = {"_at_least_one_prompt": True}
         for id, call in self._prompt_params_functions.items():

--- a/test/prompt/test_prompt_template.py
+++ b/test/prompt/test_prompt_template.py
@@ -36,6 +36,27 @@ def test_prompt_templates():
 
 
 @pytest.mark.unit
+def test_missing_prompt_template_params():
+    p = PromptTemplate("missing_params", "Here is some fake template with variable {foo} and {bar}")
+
+    # both params provided - ok
+    p.prepare(foo="foo", bar="bar")
+
+    # missing one param
+    with pytest.raises(ValueError, match=".*parameters \['bar', 'foo'\] to be provided but got only \['foo'\].*"):
+        p.prepare(foo="foo")
+
+    # missing both params
+    with pytest.raises(
+        ValueError, match=".*parameters \['bar', 'foo'\] to be provided but got none of these parameters.*"
+    ):
+        p.prepare(lets="go")
+
+    # more than both params provided - also ok
+    p.prepare(foo="foo", bar="bar", lets="go")
+
+
+@pytest.mark.unit
 def test_prompt_template_repr():
     p = PromptTemplate("t", "Here is variable {baz}")
     desired_repr = "PromptTemplate(name=t, prompt_text=Here is variable {baz}, prompt_params=['baz'])"

--- a/test/prompt/test_prompt_template.py
+++ b/test/prompt/test_prompt_template.py
@@ -37,23 +37,23 @@ def test_prompt_templates():
 
 @pytest.mark.unit
 def test_missing_prompt_template_params():
-    p = PromptTemplate("missing_params", "Here is some fake template with variable {foo} and {bar}")
+    template = PromptTemplate("missing_params", "Here is some fake template with variable {foo} and {bar}")
 
     # both params provided - ok
-    p.prepare(foo="foo", bar="bar")
+    template.prepare(foo="foo", bar="bar")
 
     # missing one param
-    with pytest.raises(ValueError, match=".*parameters \['bar', 'foo'\] to be provided but got only \['foo'\].*"):
-        p.prepare(foo="foo")
+    with pytest.raises(ValueError, match=r".*parameters \['bar', 'foo'\] to be provided but got only \['foo'\].*"):
+        template.prepare(foo="foo")
 
     # missing both params
     with pytest.raises(
-        ValueError, match=".*parameters \['bar', 'foo'\] to be provided but got none of these parameters.*"
+        ValueError, match=r".*parameters \['bar', 'foo'\] to be provided but got none of these parameters.*"
     ):
-        p.prepare(lets="go")
+        template.prepare(lets="go")
 
     # more than both params provided - also ok
-    p.prepare(foo="foo", bar="bar", lets="go")
+    template.prepare(foo="foo", bar="bar", lets="go")
 
 
 @pytest.mark.unit

